src/NodaTime/Calendars/WeekYearRule.cs

```diff
@@ -0,0 +1,199 @@
+    /// and <see cref="GetWeekOfWeekYear(LocalDate)"/> allow the week-year and week to be obtained, and
+    /// <see cref="CreateLocalDate(int, int, IsoDayOfWeek, CalendarSystem)"/> allows the reverse mapping. Note that
+    /// the calendar system does not need to be specified in the former methods as a <see cref="LocalDate"/> already
+    /// contains calendar information, and there is no method to obtain the day-of-week as that is not affected by the
+    /// week year rule being used.
+    /// </para>
+    /// <para>Specific rules are obtained using the static factory methods within this class, or with the
+    /// <see cref="Iso"/> property.</para>
+    /// <para>
+    /// All implementations within Noda Time are immutable, and it is advised that any external implementations
+    /// should be immutable too.
+    /// </para>
+    /// </remarks>
+    [Immutable]
+    public abstract class WeekYearRule
+    {
+        private static readonly WeekYearRule[] isoBasedRules = Enumerable.Range(1, 7).Select(x => new IsoBasedWeekYearRule(x)).ToArray();
+
+        /// <summary>
+        /// Returns a <see cref="WeekYearRule"/> similar to ISO-8601, but allowing the minimum number of
+        /// days in the first week to vary from the value of 4 specified by ISO-8601.
+        /// </summary>
+        /// <remarks>
+        /// <paramref name="minDaysInFirstWeek"/> determines when the first week of the week-year starts.
+        /// For any given calendar year X, consider the Monday-to-Sunday week that includes the first day of the
+        /// calendar year. Usually, some days of that week are in calendar year X, and some are in calendar year 
+        /// X-1. If <paramref name="minDaysInFirstWeek"/> or more of the days are in year X, then the week is
+        /// deemed to be the first week of week-year X. Otherwise, the week is deemed to be the last week of
+        /// week-year X-1, and the first week of week-year X starts on the following Monday.
+        /// </remarks>
+        /// <param name="minDaysInFirstWeek">The minimum number of days in the first Monday-to-Sunday week
+        /// which have to be in the new calendar year for that week to count as being in that week-year.
+        /// Must be in the range 1 to 7 inclusive.
+        /// </param>
+        /// <returns>A <see cref="WeekYearRule"/> with the specified minimum number of days in the first
+        /// week.</returns>
+        public static WeekYearRule ForMinDaysInFirstWeek(int minDaysInFirstWeek)
+        {
+            Preconditions.CheckArgumentRange(nameof(minDaysInFirstWeek), minDaysInFirstWeek, 1, 7);
+            return isoBasedRules[minDaysInFirstWeek - 1];
+        }
+
+        /// <summary>
+        /// Returns a <see cref="WeekYearRule"/> consistent with ISO-8601.
+        /// </summary>
+        /// <remarks></remarks>
+        /// <para>
+        /// In the standard ISO-8601 week algorithm, the first week of the year
```

Comment:
Again, probably want to mention weeks starting on a Monday by (ISO's) definition here.